### ISSUE TITLE
Utilizing bat and fzf environment variables

### ISSUE THREAD

**Issue by ldelossa** [ibhagwan/fzf-lua#47]
Hello, 

I manage my own vim theme with a dark and light theme. 

This works very nicely with fzf.vim when I utilize environment variables. 

```
function! DarkMode()
    let $BAT_THEME="1337"
    let $FZF_DEFAULT_OPTS='--color=dark --bind ctrl-d:preview-page-down,ctrl-u:preview-page-up'
    colorscheme vimdark
endfunction

function! LightMode()
    let $BAT_THEME="gruvbox-light"
    let $FZF_DEFAULT_OPTS='--color=light --bind ctrl-d:preview-page-down,ctrl-u:preview-page-up'
    colorscheme vimlight
endfunction

" use vimdark from 9pm to 10am
if strftime("%H") >= 21 || strftime("%H") <= 9
    call DarkMode()
else
    call LightMode()
endif

nnoremap <leader>lm :call LightMode() <CR>
nnoremap <leader>dm :call DarkMode() <CR>
```

Can fzf-lua utilize these env variables dot he right thing when I switch my color schemes? currently it seems like the fzf window utilizes its own theme and doesnt pay much attention to the env variables the binaries use.

**Comment by ibhagwan** [comment]
You’re correct, the plug-in uses it’s own variables but it’s really easy to use your own variable instead, just change it as the setup function, below are the relevant options, (you can also clear out `fzf_binds` and just set `fzf_args = “vim.env.FZF_DEFAULT_OPTS”` if you wish):

```lua
require'fzf-lua'.setup {
  fzf_args            = “—color=dark”,
  fzf_binds           = {               -- fzf '--bind=' options
      'f2:toggle-preview',
      'f3:toggle-preview-wrap',
      'shift-down:preview-page-down',
      'shift-up:preview-page-up',
      'ctrl-d:half-page-down',
      'ctrl-u:half-page-up',
      'ctrl-f:page-down',
      'ctrl-b:page-up',
      'ctrl-a:toggle-all',
      'ctrl-l:clear-query',
  },
  previewers = {
    bat = {
      cmd             = "bat",
      args            = "--style=numbers,changes --color always",
      theme           = vim.env.BAT_THEME,
      config          = nil,            -- nil uses $BAT_CONFIG_PATH
    },
}
```

**Comment by ldelossa** [comment]
I thought of this, but won't this only be resolved once at startup? Meaning, changing the theme during nvim runtime will not update the values provided into this setup method?

**Comment by ibhagwan** [comment]
Correct, if you’d like it to happen at runtime just create your own version of the function and set this right before you call it, something like this:

```lua
local my_files = function()
  vim.env.BAT_THEME = …
  require”fzf-lua”.files({ fzf_args = … })
end
```

**Comment by ldelossa** [comment]
hmm, even with providing the env variables, the popup window is being themed oddly.

**Comment by ldelossa** [comment]
![fzf-lua](https://user-images.githubusercontent.com/5642902/129758768-941d51ca-726c-40ef-a22d-8083c037b47b.png)
![fzf-vim](https://user-images.githubusercontent.com/5642902/129758770-a4ded5b6-2d49-4194-855f-070c1e90785a.png)

You can see the difference in the "buffers" screen.

**Comment by ibhagwan** [comment]
Buffers does not support highlighting in previews yet, I will add that when I add the built-in previewer with treesitter support (it’s on the main page TODO list).

If you’re talking about the border color that’s because it’s drawn by neovim and not fzf, for some reason lots of color schemes have `FloatBorder` linked to a diff highlight (I.e. not set to a specific color), I had an issue where a plug-in was changing my `FloatBorder` color so the first load would have a different color like you’re experiencing (in my case it was `lsp-signature` plug-in taking over the color), once I set it to a specific color the plug-in was no longer modifying `FloatBorder`.

You can either set `FloatBorder` to a specific color in your color scheme or relink it specifically at:

```lua
require'fzf-lua'.setup {
  winopts = {
      window_on_create = function()         -- nvim window options override
        vim.cmd("set winhl=Normal:Normal,FloatBorder:<highlight>")  -- popup bg match normal windows
      end,
    },
}
```

**Comment by ldelossa** [comment]
@ibhagwan good to know the highlighting is not an error.

I'm more referring to the text color of each buffer/file in the output. fzf.vim is keeping the blue while fzf.lua is making it green and hard to see. 

Basically, fzf.vim is just opening a termal tab, and fzf.lua is writing terminal output to native vim windows? Am I understanding that correctly?

**Comment by ibhagwan** [comment]
> I'm more referring to the text color of each buffer/file in the output. fzf.vim is keeping the blue while fzf.lua is making it green and hard to see.

The green is hard coded only for the current buffer name, I think I can remove that safely as the `—header-line` should color it anyways based on the fzf theme.

You can also avoid displaying the current buffer with `buffers({ ignore_current_buffer = true }).

> Basically, fzf.vim is just opening a termal tab, and fzf.lua is writing terminal output to native vim windows? Am I understanding that correctly?

Nope, they’re both actually very similar, `fzf-lua` also opens a popular and then a terminal window inside it (via `nvim-fzf` plug-in).

**Comment by ibhagwan** [comment]
I removed the hard-coded green color when using the header line, his should solve the highlight issue in your case, the current buffer will always be set as the header and use the header color of your current fzf theme.

As for the highlights, it's pretty easy for me to solve by using `bat` for all buffers but that creates an issue previewing buffers with no files attached yet (and also inconsistency with files that aren't yet saved to disk).

I plan to solve this for `buffers` as well as for `marks` (similar problem) soon as it's the next issue [on my TODO list](https://github.com/ibhagwan/fzf-lua#TODO).

**Comment by ldelossa** [comment]
I pulled your latest changes, but seem to be getting the same green highlighting as before.

**Comment by ibhagwan** [comment]
> I pulled your latest changes, but seem to be getting the same green highlighting as before.

The default color for fzf `—header-line` is also green, it’s now set by fzf which I believe you can control yourself, I’m not setting that color unless you send `sort_last_used = false` as an argument.

If your current buffer is not selectable that means it’s a header line and [I don’t set it’s color manually since the last change](https://github.com/ibhagwan/fzf-lua/blob/main/lua/fzf-lua/providers/buffers.lua#L98).

I can’t recall ATM - does `fzf.vim` use a header line? Is the current buffer selectable?

Perhaps you can comment L99 in buffers.lua and see if that makes a change? That’s the only palace where the color could be set.

**Comment by ibhagwan** [comment]
Just tested this in the shell, `line1` is green:

```sh
echo "line1\nline2\nline3\n" | fzf --ansi --header-lines=1
```
Once you fix that `fzf-lua` will behave the same.

**Comment by ldelossa** [comment]
@ibhagwan thanks for hanging in on this one with me. 

This is not working for me, I am now explicitly setting the header in my env var. Here is fzf-lua open and my status line just after echoing "lua print(vim.env.FZF_DEFAULT_OPTS)" 

You'll notice I'm defining the header color explicitly and that works on the CLI for me. 

![fzf-header](https://user-images.githubusercontent.com/5642902/129812120-550b8d08-dbc0-4235-bdbb-53eb1c9b68b6.png)

**Comment by ibhagwan** [comment]
It seems the current buffer is selectable, that’s in fact the only condition where I do set the green color manual, I now see that your setup has `sort_lastused = false`, change that to `true` and it will be solved :)

**Comment by ldelossa** [comment]
works! thanks a ton. 

There are still a few more green elements left, like the brackets, and file count (2/2) portion. Are those set by you or fzf?

**Comment by ibhagwan** [comment]
> works! thanks a ton.
> 
> There are still a few more green elements left, like the brackets, and file count (2/2) portion. Are those set by you or fzf?

The green brackets are by me (can remove if that’s a problem lmk), the file count is internal to fzf.

**Comment by ibhagwan** [comment]
I just gave the `[]` the same treatment as the buffer name (i.e. do not color if using header line with `sort_lastused = true`), the only colors left hard-coded are the yellow for buffer ID (inside the brackets) and red for the buffer flag (`%`).

**Comment by ldelossa** [comment]
Yellow buffer ID is not very visible, anyway to allow this to be customized as well?

**Comment by ibhagwan** [comment]
> Yellow buffer ID is not very visible, anyway to allow this to be customized as well?

I don’t control the colors directly, it’s your terminal’s own yellow, that depends on your terminal settings (kitty, alacritty, iterm, etc). It’s also possibly set by your nvim color scheme, usually like this:

```vim
let g:terminal_color_0 = s:bg_bright.gui
let g:terminal_color_1 = s:red.gui
let g:terminal_color_2 = s:green.gui
let g:terminal_color_3 = s:yellow.gui
let g:terminal_color_4 = s:blue.gui
let g:terminal_color_5 = s:purple.gui
let g:terminal_color_6 = s:cyan.gui
let g:terminal_color_7 = s:norm.gui
let g:terminal_color_8 = s:bg_bright.gui
let g:terminal_color_9 = s:dark_red.gui
let g:terminal_color_10 = s:dark_green.gui
let g:terminal_color_11 = s:dark_yellow.gui
let g:terminal_color_12 = s:dark_blue.gui
let g:terminal_color_13 = s:dark_purple.gui
let g:terminal_color_14 = s:dark_cyan.gui
let g:terminal_color_15 = s:norm_subtle.gui
```

**Comment by ldelossa** [comment]
> > Yellow buffer ID is not very visible, anyway to allow this to be customized as well?
> 
> I don’t control the colors directly, it’s your terminal’s own yellow, that depends on your terminal settings (kitty, alacritty, iterm, etc). It’s also possibly set by your nvim color scheme, usually like this:
> 
> ```viml
> let g:terminal_color_0 = s:bg_bright.gui
> let g:terminal_color_1 = s:red.gui
> let g:terminal_color_2 = s:green.gui
> let g:terminal_color_3 = s:yellow.gui
> let g:terminal_color_4 = s:blue.gui
> let g:terminal_color_5 = s:purple.gui
> let g:terminal_color_6 = s:cyan.gui
> let g:terminal_color_7 = s:norm.gui
> let g:terminal_color_8 = s:bg_bright.gui
> let g:terminal_color_9 = s:dark_red.gui
> let g:terminal_color_10 = s:dark_green.gui
> let g:terminal_color_11 = s:dark_yellow.gui
> let g:terminal_color_12 = s:dark_blue.gui
> let g:terminal_color_13 = s:dark_purple.gui
> let g:terminal_color_14 = s:dark_cyan.gui
> let g:terminal_color_15 = s:norm_subtle.gui
> ```

Thanks a ton. Im cool with the issue being closed

**Comment by ibhagwan** [comment]
Ty! Glad we could get to the bottom of this :)

**Comment by ibhagwan** [comment]
FYI, today I merged PR #69 which added support for fzf’s `—color` option and you can now set fzf’s colors based on neovim’s highlights from your scheme, checkout `fzf_colors` in the README to see how to configure it.
